File: cpp/cpp_1000-/1302.cc
// C++
// leetcode 1302
// https://leetcode.cn/problems/deepest-leaves-sum/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
    	int sum = 0;
    	queue<TreeNode*> q;
        q.push(root);
    	TreeNode* p;
    	while (!q.empty()) {
    		int sz = q.size();
    		sum = 0;
    		while (sz--) {
    			p = q.front();
    			sum += p->val;
    			q.pop();
    			if (p -> left)
    				q.push(p->left);
    			if (p -> right)
    				q.push(p->right);
    		}
    	}
    	return sum;
    }
};

File: Go/Go_1000-/1302.go
// Go
// leetcode 1302
// https://leetcode.cn/problems/deepest-leaves-sum/

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func deepestLeavesSum(root *TreeNode) int {
	q := []*TreeNode{root}
	var sum int
	for len(q) != 0 {
		sz := len(q)
		sum = 0
		for i := 0; i < sz; i++ {
			sum += q[i].Val
			if q[i].Left != nil {
				q = append(q, q[i].Left)
			}
			if q[i].Right != nil {
				q = append(q, q[i].Right)
			}
		}
		q = q[sz:]
	}
	return sum
}